### PR TITLE
Add report split cmd support.

### PR DESCRIPTION
**Pull request by qiuyesuifeng**
@siddontang @disksing @shenli  PTAL

### CODE REVIEW

File: server/cluster_worker.go [hunk 1]
```diff
@@ -113,3 +115,41 @@ func (c *RaftCluster) handleAskSplit(request *pdpb.AskSplitRequest) (*pdpb.AskSp
...
+		log.Warnf("report split region is invalid - %v, %v", request, errors.ErrorStack(err))
+		return nil, errors.Trace(err)
+	}
+
+	// Build origin region by using left and right.
+	originRegion := cloneRegion(left)
+	originRegion.RegionEpoch = nil
+	originRegion.EndKey = right.GetEndKey()
+
+	// Wrap report split message as an Operator, and added into cache.
```
Comment (shenli) [inline server/cluster_worker.go]: added -> add

Comment (qiuyesuifeng) [inline server/cluster_worker.go]: Gotcha.

File: server/cluster_worker.go [hunk 2]
```diff
@@ -113,3 +115,41 @@ func (c *RaftCluster) handleAskSplit(request *pdpb.AskSplitRequest) (*pdpb.AskSp
...
+	op := newSplitOperator(originRegion, left, right)
+	c.balancerWorker.historyOperators.add(originRegion.GetId(), op)
+
+	return &pdpb.ReportSplitResponse{}, nil
```
Comment (shenli) [inline server/cluster_worker.go]: Return an empty response?

File: server/cluster_worker.go [hunk 3]
```diff
@@ -113,3 +115,41 @@ func (c *RaftCluster) handleAskSplit(request *pdpb.AskSplitRequest) (*pdpb.AskSp
...
+func (c *RaftCluster) checkSplitRegion(left *metapb.Region, right *metapb.Region) error {
+	if left == nil || right == nil {
+		return errors.New("invalid split region")
+	}
+
+	if !bytes.Equal(left.GetEndKey(), right.GetStartKey()) {
+		return errors.New("invalid split region")
+	}
+
+	if bytes.Compare(left.GetStartKey(), right.GetEndKey()) >= 0 {
```
Comment (disksing) [inline server/cluster_worker.go]: Is there any chance that `right.GetEndKey() == nil`?

File: server/cluster_worker.go (hunk 2 above)
Comment (qiuyesuifeng) [inline server/cluster_worker.go]: Yes, it is only a report from tikv, like store/region heartbeat, there is no need to return a detail response.

File: server/cluster_worker.go (hunk 3 above)
Comment (qiuyesuifeng) [inline server/cluster_worker.go]: My fault, i forgot it.

File: server/cluster_worker.go [hunk 4]
```diff
@@ -113,3 +115,41 @@ func (c *RaftCluster) handleAskSplit(request *pdpb.AskSplitRequest) (*pdpb.AskSp
...
+	right := request.GetRight()
+
+	err := c.checkSplitRegion(left, right)
+	if err != nil {
+		log.Warnf("report split region is invalid - %v, %v", request, errors.ErrorStack(err))
+		return nil, errors.Trace(err)
+	}
+
+	// Build origin region by using left and right.
+	originRegion := cloneRegion(left)
```
Comment (siddontang) [inline server/cluster_worker.go]: do we really need origin?

Comment (qiuyesuifeng) [inline server/cluster_worker.go]: It is better for web UI to display the region split event.

**Comment by disksing** [comment]
LGTM

File: server/cluster_worker.go (hunk 4 above)
Comment (siddontang) [inline server/cluster_worker.go]: the epoch is nil, does web UI not care this?

Comment (qiuyesuifeng) [inline server/cluster_worker.go]: It will not use epoch i think, it only cares the origin region and split regions info, like starkey/endkey and peers.

Comment (qiuyesuifeng) [inline server/cluster_worker.go]: I only don't want the epoch value to make someone confused, so i make it as a nil value.

Comment (siddontang) [inline server/cluster_worker.go]: you can use left version - 1 to construct it.

Comment (siddontang) [inline server/cluster_worker.go]: but if you don't care epoch, nil is ok. Of course, UI should not show left and right epoch too.

Comment (qiuyesuifeng) [inline server/cluster_worker.go]: Yes, the epoch value is useless.

**Comment by siddontang** [comment]
LGTM

**Pull request merged**